### PR TITLE
Ensure proper istio RBAC permissions when `ManagedIstio` feature gate is enabled by default

### PR DESCRIPTION
**Pull request by amshuman-kr**
**How to categorize this PR?**

/area security 
/kind impediment

**What this PR does / why we need it**:
I found some RBAC issues while bootstrapping a seed from scratch (with gardenlet installed on it) in the gardener release `v1.28.1`. I think the same issues exist in the master branch as well. 

```
level=error msg="Seed bootstrapping failed: failed to apply manifests: 4 errors occurred: [could not apply object of kind \"DestinationRule\" \"istio-system/default\": destinationrules.networking.istio.io \ "default\" is forbidden: User \"system:serviceaccount:garden:gardenlet\" cannot get resource \"destinationrules\" in API group \"networking.istio.io\" in the namespace \"istio-system\", could not apply object of kind \"PeerAuthentication\" \"istio-system/default\": peerauthentications.security.istio.io \"default\" is forbidden: User \"system:serviceaccount:garden:gardenlet\" cannot get resource \"peerauthentications\" in API group \"security.istio.io\" in the namespace \"istio-system\", could not apply object of kind \"Sidecar\" \"istio-system/default\": sidecars.networking.istio.io \"default\" is forbidden: User \"system:serviceaccount:garden:gardenlet\" cannot get resource \"sidecars\" in API group \"networking.istio.io\" in the namespace \"istio-system\", could not apply object of kind \"ValidatingWebhookConfiguration\" \"/istiod\": validatingwebhookconfigurations.admissionregistration.k8s.io \"istiod\" is forbidden: User \"system:serviceaccount:garden:gardenlet\" cannot get resource \"validatingwebhookconfigurations\" in API group \"admissionregistration.k8s.io\" at the cluster scope]" seed=test-aws-seed
```

Ensure proper istio RBAC permissions when `ManagedIstio` feature gate is enabled by default.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

Only the commit https://github.com/gardener/gardener/pull/4481/commits/aed80ede310bfaf0ab7dbaa6e30d61739545d017 is relevant.

 I have based this PR on #4494 because I wanted to add the template definition `gardenlet.managed-istio-enabled` in the file `charts/gardener/gardenlet/charts/runtime/templates/_feature-gates.tpl` which was introduced in #4494.

I will rebase this PR after #4494 is merged.

**Release note**:

```other operator
Ensure proper istio RBAC permissions when `ManagedIstio` feature gate is enabled by default.
```

### CODE REVIEW

**Comment by vpnachev** [comment]
Which gardenlet feature gates are enabled and which disabled?

**Comment by amshuman-kr** [comment]
> Which gardenlet feature gates are enabled and which disabled?

I was using the default configuration as in the gardenlet helm chart which seems to have the [featureGates configuration as empty](https://github.com/gardener/gardener/blob/master/charts/gardener/gardenlet/values.yaml#L135).

**Comment by vpnachev** [comment]
On start-up gardenlet log a message with the enabled and disabled feature gates, please share it.

**Comment by amshuman-kr** [comment]
> On start-up gardenlet log a message with the enabled and disabled feature gates, please share it.

```
ime="2021-08-09T05:29:43Z" level=info msg="Starting Gardenlet..."
time="2021-08-09T05:29:43Z" level=info msg="Feature Gates: "
```

**Comment by vpnachev** [comment]
/assign

I am trying to reproduce with empty featureGates spec for gardenlet.

**Comment by timuthy** [comment]
A related error I noticed when running with:
```
  ManagedIstio: false
  APIServerSNI: true
```

```
time="2021-08-09T15:43:36+02:00" level=error msg="Seed bootstrapping failed: failed to apply manifests: 3 errors occurred: [could not apply object of kind \"EnvoyFilter\" \"istio-ingress/proxy-protocol\": envoyfilters.networking.istio.io \"proxy-protocol\" is forbidden: User \"system:serviceaccount:garden:gardenlet\" cannot get resource \"envoyfilters\" in API group \"networking.istio.io\" in the namespace \"istio-ingress\", could not apply object of kind \"Gateway\" \"istio-ingress/proxy-protocol\": gateways.networking.istio.io \"proxy-protocol\" is forbidden: User \"system:serviceaccount:garden:gardenlet\" cannot get resource \"gateways\" in API group \"networking.istio.io\" in the namespace \"istio-ingress\", could not apply object of kind \"VirtualService\" \"istio-ingress/proxy-protocol-blackhole\": virtualservices.networking.istio.io \"proxy-protocol-blackhole\" is forbidden: User \"system:serviceaccount:garden:gardenlet\" cannot get resource \"virtualservices\" in API group \"networking.istio.io\" in the namespace \"istio-ingress\"]" seed=gcp
```

**Review by vpnachev — CHANGES_REQUESTED** [review]
All required permissions for the gardenlet are already provided via https://github.com/gardener/gardener/blob/master/charts/gardener/gardenlet/charts/runtime/templates/clusterrole-managed-istio.yaml.

In your case, you are deploying the gardenlet chart with `.global.gardenlet.config.featureGates={}` and this way the clusterrole(binding) `gardener.cloud:system:gardenlet:managed-istio` is not deployed. However, the `APIServerSNI` and `ManagedIstio` gardenlet feature gates are enabled by default and due to this mismatch, the gardenlet is not authorized with the required permissions.



If you really want to fix this, you have to improve this check https://github.com/gardener/gardener/blob/768a2a1fa9639a2b0524a00fdd1b677318d07431/charts/gardener/gardenlet/charts/runtime/templates/clusterrole-managed-istio.yaml#L1-L3 to also consider `ManagedIstio` as enabled by default.

**Comment by vpnachev** [comment]
> A related error I noticed when running with:
> 
> ```
>   ManagedIstio: false
>   APIServerSNI: true
> ```
> 
> ```
> time="2021-08-09T15:43:36+02:00" level=error msg="Seed bootstrapping failed: failed to apply manifests: 3 errors occurred: [could not apply object of kind \"EnvoyFilter\" \"istio-ingress/proxy-protocol\": envoyfilters.networking.istio.io \"proxy-protocol\" is forbidden: User \"system:serviceaccount:garden:gardenlet\" cannot get resource \"envoyfilters\" in API group \"networking.istio.io\" in the namespace \"istio-ingress\", could not apply object of kind \"Gateway\" \"istio-ingress/proxy-protocol\": gateways.networking.istio.io \"proxy-protocol\" is forbidden: User \"system:serviceaccount:garden:gardenlet\" cannot get resource \"gateways\" in API group \"networking.istio.io\" in the namespace \"istio-ingress\", could not apply object of kind \"VirtualService\" \"istio-ingress/proxy-protocol-blackhole\": virtualservices.networking.istio.io \"proxy-protocol-blackhole\" is forbidden: User \"system:serviceaccount:garden:gardenlet\" cannot get resource \"virtualservices\" in API group \"networking.istio.io\" in the namespace \"istio-ingress\"]" seed=gcp
> ```

I would say this expected as `APIServerSNI` do require Istio to be installed in the seed cluster and it was done via external to the Gardener component, then I also expect this external component to authorize gardenlet to use it.

**Comment by timuthy** [comment]
> I would say this expected as APIServerSNI do require Istio to be installed in the seed cluster and it was done via external to the Gardener component, then I also expect this external component to authorize gardenlet to use it.

Only Gardenlet knows what permissions are required to deploy `APIServerSNI`, so I don't think that this is the external component's responsibility to find that out. And as these permissions evolve, they should also be maintained with the Gardenlet chart.

**Comment by amshuman-kr** [comment]
> f you really want to fix this, you have to improve this check
> 
> https://github.com/gardener/gardener/blob/768a2a1fa9639a2b0524a00fdd1b677318d07431/charts/gardener/gardenlet/charts/runtime/templates/clusterrole-managed-istio.yaml#L1-L3
> 
> to also consider `ManagedIstio` as enabled by default.

@vpnachev Thanks a lot for finding the root cause and the suggestion! I can see that the feature gates defaults are initialised [here](https://github.com/gardener/gardener/blob/e5b22bcc58c9a4205aa0e88bd3fc352d7bf05dfc/pkg/gardenlet/features/features.go#L27-L38). But I can't think of a good way of re-using this in the helm charts. And duplicating the defaulting in helm chart doesn't look good either. Any suggestions?

**Comment by vpnachev** [comment]
> A related error I noticed when running with:
> 
> ```
>   ManagedIstio: false
>   APIServerSNI: true
> ```
> 
> ```
> time="2021-08-09T15:43:36+02:00" level=error msg="Seed bootstrapping failed: failed to apply manifests: 3 errors occurred: [could not apply object of kind \"EnvoyFilter\" \"istio-ingress/proxy-protocol\": envoyfilters.networking.istio.io \"proxy-protocol\" is forbidden: User \"system:serviceaccount:garden:gardenlet\" cannot get resource \"envoyfilters\" in API group \"networking.istio.io\" in the namespace \"istio-ingress\", could not apply object of kind \"Gateway\" \"istio-ingress/proxy-protocol\": gateways.networking.istio.io \"proxy-protocol\" is forbidden: User \"system:serviceaccount:garden:gardenlet\" cannot get resource \"gateways\" in API group \"networking.istio.io\" in the namespace \"istio-ingress\", could not apply object of kind \"VirtualService\" \"istio-ingress/proxy-protocol-blackhole\": virtualservices.networking.istio.io \"proxy-protocol-blackhole\" is forbidden: User \"system:serviceaccount:garden:gardenlet\" cannot get resource \"virtualservices\" in API group \"networking.istio.io\" in the namespace \"istio-ingress\"]" seed=gcp
> ```

I took a deeper look into this. I believe it is due to https://github.com/gardener/gardener/blob/1ab6be7c84ba36c10a799c2e07ac07a97ccda912/pkg/operation/seed/seed.go#L752-L786 introduced with #4117.

I.e. resource from https://github.com/gardener/gardener/tree/master/charts/istio/istio-proxy-protocol are deployed/deleted depending only to the `APIServerSNI` feature gate. So, I would say the findings of @timuthy is different problem and permissions for this `proxy-protocol` chart resources needs to be added to https://github.com/gardener/gardener/blob/master/charts/gardener/gardenlet/charts/runtime/templates/clusterrole-gardenlet.yaml.


> But I can't think of a good way of re-using this in the helm charts. And duplicating the defaulting in helm chart doesn't look good either. Any suggestions?

@amshuman-kr I have no better suggestion, but for sure providing gardenlet with permissions that it does not need is not a better solution. You can try to achieve `enabled by default` via something like
```yaml
{{- define "gardenlet.managed-istio-enabled" }}
{{- if .Values.global.gardenlet.config.featureGates }}
{{- if hasKey .Values.global.gardenlet.config.featureGates "ManagedIstio" }}
{{ .Values.global.gardenlet.config.featureGates.ManagedIstio }}
{{- else }}
true
{{- end }}
{{- else }}
true
{{- end }}
{{- end -}}
```

**Comment by amshuman-kr** [comment]
> but for sure providing gardenlet with permissions that it does not need is not a better solution.

Agreed. I will undo the current changes in the PR.

> You can try to achieve enabled by default via something like

```
{{- define "gardenlet.managed-istio-enabled" }}
{{- if .Values.global.gardenlet.config.featureGates }}
{{- if hasKey .Values.global.gardenlet.config.featureGates "ManagedIstio" }}
{{ .Values.global.gardenlet.config.featureGates.ManagedIstio }}
{{- else }}
true
{{- end }}
{{- else }}
true
{{- end }}
{{- end -}}
```

This is what I meant by _duplication of defaulting_. If the defaults change then it might have to be changed in two places. If this is OK, I will make change (along with undoing the current changes).

**Comment by vpnachev** [comment]
> This is what I meant by duplication of defaulting. If the defaults change then it might have to be changed in two places. If this is OK, I will make change (along with undoing the current changes).

I don't see how it can be avoided. The default must be defined in the golang code as gardenlet can be deployed via other means. The chart does not necessary have access to the source code, so it cannot evaluate during rendering time if a feature gate is enabled or not. The reverse is also valid, the golang code does  not necessarily have access to the charts to initialize the defaulting from there. Therefore I don't think duplication can be avoided. I might be wrong and missing a solution how it can be implemented only once, do you have any ideas?

**Comment by amshuman-kr** [comment]
> Therefore I don't think duplication can be avoided. I might be wrong and missing a solution how it can be implemented only once, do you have any ideas?

Yes. I don't see any way how the duplication can be avoided. I will make the change as you suggested.

**Comment by amshuman-kr** [comment]
@vpnachev I thought that it might be simpler to default the feature gates in the `values.yaml`. I have made the changes. Can you PTAL?

**Comment by vpnachev** [comment]
> @vpnachev I thought that it might be simpler to default the feature gates in the `values.yaml`. I have made the changes. Can you PTAL?

Sorry, but this does not work as different values file can be provided and there this defaulting might not be reflected.
Please take a look at https://github.com/gardener/gardener/pull/4494/commits/077071b57081ace742af7157d4e87d19602c6589 where I propose same change but for the `APIServerSNI` feature gate.

**Comment by amshuman-kr** [comment]
> Sorry, but this does not work as different values file can be provided and there this defaulting might not be reflected.
> Please take a look at [077071b](https://github.com/gardener/gardener/commit/077071b57081ace742af7157d4e87d19602c6589) where I propose same change but for the `APIServerSNI` feature gate.

I have made the changes as you suggested. I have based this PR on #4494 because I wanted to add the template definition `gardenlet.managed-istio-enabled` in the file `charts/gardener/gardenlet/charts/runtime/templates/_feature-gates.tpl` which was introduced in #4494. 

I will rebase this PR after #4494 is merged.

**Comment by amshuman-kr** [comment]
> /needs rebase

Done in https://github.com/gardener/gardener/pull/4481#event-5166774000

**Review by rfranzke — DISMISSED** [review]
/lgtm

File: charts/gardener/gardenlet/charts/runtime/templates/_feature-gates.tpl [hunk 1]
```diff
@@ -9,3 +9,15 @@ true
...
+{{- if .Values.global.gardenlet.config.featureGates }}
+{{- if hasKey .Values.global.gardenlet.config.featureGates "ManagedIstio" }}
+{{ .Values.global.gardenlet.config.featureGates.ManagedIstio }}
+{{- else }}
+true
+{{- end }}
+{{- else }}
+true
+{{- end }}
+{{- end -}}
```
Comment (vpnachev) [inline charts/gardener/gardenlet/charts/runtime/templates/_feature-gates.tpl]:
```suggestion
{{- define "gardenlet.managed-istio-enabled" -}}
{{- if .Values.global.gardenlet.config.featureGates -}}
{{- if hasKey .Values.global.gardenlet.config.featureGates "ManagedIstio" -}}
{{- .Values.global.gardenlet.config.featureGates.ManagedIstio -}}
{{- else -}}
true
{{- end -}}
{{- else -}}
true
{{- end -}}
{{- end -}}
```

**Review by vpnachev — CHANGES_REQUESTED** [review]
You need to trim the leading and trailing white spaces in the template.

**Comment by amshuman-kr** [comment]
> You need to trim the leading and trailing white spaces in the template.

Thanks @vpnachev. I have applied your suggestions.

**Review by vpnachev — APPROVED** [review]
/lgtm
/squash

**Pull request merged**